bundle/artifacts/artifacts.go

```diff
@@ -90,38 +90,52 @@ func (m *basicUpload) Apply(ctx context.Context, b *bundle.Bundle) error {
 		return fmt.Errorf("artifact doesn't exist: %s", m.name)
 	}
 
-	if artifact.File == "" {
+	if len(artifact.Files) == 0 {
 		return fmt.Errorf("artifact source is not configured: %s", m.name)
 	}
 
-	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Upload(%s): Uploading...", m.name))
-
-	r, err := uploadArtifact(ctx, artifact, b)
+	err := uploadArtifact(ctx, artifact, b)
 	if err != nil {
 		return fmt.Errorf("artifacts.Upload(%s): %w", m.name, err)
 	}
 
-	artifact.RemotePath = r
-	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Upload(%s): Upload succeeded", m.name))
 	return nil
 }
 
-// Function to upload artifact as a library to Workspace
-// Currenly it does not work correctly because Workspace.Import API can not import libraries
-// but only notebooks or files
-func uploadArtifact(ctx context.Context, a *config.Artifact, b *bundle.Bundle) (string, error) {
-	raw, err := os.ReadFile(a.File)
+func uploadArtifact(ctx context.Context, a *config.Artifact, b *bundle.Bundle) error {
+	for i := range a.Files {
+		f := &a.Files[i]
+		if f.Libraries != nil {
```

Comment:
Why check for `f.Libraries` but not use it in the below branch?

Reply:
This is clear to me after reading the whole change, but can use a few more comments. The part where it matches to task libraries and then links that back into artifacts is not super clear. It's an indicator that it should be uploaded, but that can be made more explicit, IMO. E.g. add a `NeedsUpload` helper that checks for `f.Libraries != nil` with a bunch of comments explaining the logic.